File: worker/src/services.ts
// Общая бизнес-логика: выплаты, возвраты, уведомления сторон (порт bot/services.py).

import type { Ctx } from "./ctx";
import * as d from "./db";
import type { DealRow } from "./db";
import { CryptoPayError } from "./cryptopay";
import { buyerEscrowKb, rateKb, sellerEscrowKb } from "./keyboards";
import type { InlineKeyboardMarkup } from "./types";
import { fmtAmount, round8 } from "./utils";

export function payoutAmount(amount: number, commissionPercent: number): number {
  return round8(amount * (1 - commissionPercent / 100));
}

/** Отправляет личное сообщение. false — если пользователь не запускал бота. */
export async function notify(
  ctx: Ctx,
  userId: number | null,
  text: string,
  replyMarkup?: InlineKeyboardMarkup,
): Promise<boolean> {
  if (userId === null) return false;
  try {
    await ctx.tg.sendMessage(userId, text, { reply_markup: replyMarkup });
    return true;
  } catch {
    console.warn(`Не удалось отправить сообщение пользователю ${userId}`);
    return false;
  }
}

export async function askRatings(ctx: Ctx, deal: DealRow): Promise<void> {
  const text =
    `⭐️ Оцените вашего контрагента по сделке <b>#${deal.id}</b>.\n` +
    "Оценка попадёт в его репутацию.";
  for (const uid of [deal.seller_id, deal.buyer_id]) {
    await notify(ctx, uid, text, rateKb(deal.id));
  }
}

/** Выплата продавцу из холда. initiator — 'buyer' или 'admin'. */
export async function releaseToSeller(
  ctx: Ctx,
  deal: DealRow,
  initiator: "buyer" | "admin",
): Promise<boolean> {
  const amount = payoutAmount(deal.amount, ctx.cfg.commissionPercent);
  try {
    // Без comment: Crypto Pay запрещает комментарии к небольшим переводам
    // (CANNOT_ATTACH_COMMENT), а стороны и так получают уведомление от бота.
    await ctx.cp.transfer({
      user_id: deal.seller_id!,
      asset: deal.asset,
      amount,
      spend_id: `payout_${deal.id}`,
    });
  } catch (e) {
    const name = e instanceof CryptoPayError ? e.errorName : String(e);
    console.error(`Ошибка выплаты по сделке ${deal.id}: ${name}`);
    for (const adminId of ctx.cfg.adminIds) {
      await notify(ctx, adminId, `🚨 Ошибка выплаты продавцу по сделке #${deal.id}: ${name}`);
    }
    return false;
  }

  await ctx.db.setStatus(deal.id, d.COMPLETED);
  const who =
    initiator === "buyer" ? "Покупатель подтвердил получение" : "Спор решён администратором";
  await notify(
    ctx,
    deal.seller_id,
    `✅ ${who}. Вам выплачено <b>${fmtAmount(amount)} ${deal.asset}</b> ` +
      `по сделке #${deal.id} (комиссия сервиса ${ctx.cfg.commissionPercent}%).\n` +
      "Средства зачислены на ваш баланс в @CryptoBot.",
  );
  await notify(
    ctx,
    deal.buyer_id,
    `✅ Сделка #${deal.id} завершена. Средства выплачены продавцу.`,
  );
  const fresh = await ctx.db.getDeal(deal.id);
  if (fresh) await askRatings(ctx, fresh);
  return true;
}

/** Возврат покупателю из холда. initiator — 'seller' или 'admin'. */
export async function refundToBuyer(
  ctx: Ctx,
  deal: DealRow,
  initiator: "seller" | "admin",
): Promise<boolean> {
  try {
    await ctx.cp.transfer({
      user_id: deal.buyer_id!,
      asset: deal.asset,
      amount: deal.amount,
      spend_id: `refund_${deal.id}`,
    });
  } catch (e) {
    const name = e instanceof CryptoPayError ? e.errorName : String(e);
    console.error(`Ошибка возврата по сделке ${deal.id}: ${name}`);
    for (const adminId of ctx.cfg.adminIds) {
      await notify(ctx, adminId, `🚨 Ошибка возврата покупателю по сделке #${deal.id}: ${name}`);
    }
    return false;
  }

  await ctx.db.setStatus(deal.id, d.REFUNDED);
  const who = initiator === "seller" ? "Продавец отменил сделку" : "Спор решён администратором";
  await notify(
    ctx,
    deal.buyer_id,
    `↩️ ${who}. Вам возвращено <b>${fmtAmount(deal.amount)} ${deal.asset}</b> ` +
      `по сделке #${deal.id}.\n` +
      "Средства зачислены на ваш баланс в @CryptoBot.",
  );
  await notify(
    ctx,
    deal.seller_id,
    `↩️ Сделка #${deal.id} закрыта, средства возвращены покупателю.`,
  );
  return true;
}

/**
 * Перевод оплаченной сделки в холд + уведомление сторон
 * (порт bot/watcher.py; вызывается из вебхука Crypto Pay и из cron-фолбэка).
 */
export async function markDealPaid(ctx: Ctx, dealId: string): Promise<void> {
  // Перепроверяем статус, чтобы не обработать дважды
  const deal = await ctx.db.getDeal(dealId);
  if (!deal || deal.status !== d.WAITING_PAYMENT) return;

  await ctx.db.setStatus(deal.id, d.PAID);
  console.log(`Сделка ${deal.id} оплачена, средства в холде`);

  const amount = `${fmtAmount(deal.amount)} ${deal.asset}`;
  await notify(
    ctx,
    deal.buyer_id,
    `🔒 Оплата получена! <b>${amount}</b> по сделке #${deal.id} ` +
      "заморожены у гаранта.\n\n" +
      "Когда получите товар/услугу — подтвердите это кнопкой ниже, " +
      "и деньги уйдут продавцу.",
    buyerEscrowKb(deal.id),
  );
  await notify(
    ctx,
    deal.seller_id,
    `🔒 Покупатель оплатил сделку #${deal.id} — <b>${amount}</b> в холде у гаранта.\n\n` +
      "Можете передавать товар/услугу. Деньги будут выплачены вам, " +
      "как только покупатель подтвердит получение.",
    sellerEscrowKb(deal.id),
  );
}

/** Cron-фолбэк: опрашивает Crypto Pay по всем сделкам, ждущим оплату. */
export async function checkPendingPayments(ctx: Ctx): Promise<void> {
  const deals = await ctx.db.dealsWaitingPayment();
  if (!deals.length) return;

  const byInvoice = new Map(deals.map((deal) => [deal.invoice_id!, deal]));
  const invoices = await ctx.cp.getInvoices([...byInvoice.keys()]);

  for (const invoice of invoices) {
    if (invoice.status !== "paid") continue;
    const deal = byInvoice.get(invoice.invoice_id);
    if (deal) await markDealPaid(ctx, deal.id);
  }
}

File: bot/services.py
"""Общая бизнес-логика: выплаты, возвраты, уведомления сторон."""

import logging

from aiogram import Bot

from . import db as d
from .config import Config
from .cryptopay import CryptoPay, CryptoPayError
from .db import Database
from .keyboards import rate_kb
from .utils import fmt_amount

logger = logging.getLogger(__name__)


def payout_amount(amount: float, commission_percent: float) -> float:
    return round(amount * (1 - commission_percent / 100), 8)


async def notify(bot: Bot, user_id: int | None, text: str, **kwargs) -> bool:
    """Отправляет личное сообщение. False — если пользователь не запускал бота."""
    if user_id is None:
        return False
    try:
        await bot.send_message(user_id, text, **kwargs)
        return True
    except Exception:
        logger.warning("Не удалось отправить сообщение пользователю %s", user_id)
        return False


async def ask_ratings(bot: Bot, deal) -> None:
    text = (
        f"⭐️ Оцените вашего контрагента по сделке <b>#{deal['id']}</b>.\n"
        "Оценка попадёт в его репутацию."
    )
    for uid in (deal["seller_id"], deal["buyer_id"]):
        await notify(bot, uid, text, reply_markup=rate_kb(deal["id"]))


async def release_to_seller(
    bot: Bot, db: Database, cp: CryptoPay, config: Config, deal, initiator: str
) -> bool:
    """Выплата продавцу из холда. initiator — 'buyer' или 'admin'."""
    amount = payout_amount(deal["amount"], config.commission_percent)
    try:
        await cp.transfer(
            user_id=deal["seller_id"],
            asset=deal["asset"],
            amount=amount,
            spend_id=f"payout_{deal['id']}",
            comment=f"Выплата по сделке #{deal['id']}",
        )
    except CryptoPayError as e:
        logger.error("Ошибка выплаты по сделке %s: %s", deal["id"], e)
        for admin_id in config.admin_ids:
            await notify(
                bot, admin_id,
                f"🚨 Ошибка выплаты продавцу по сделке #{deal['id']}: {e.name}",
            )
        return False

    await db.set_status(deal["id"], d.COMPLETED)
    who = "Покупатель подтвердил получение" if initiator == "buyer" else "Спор решён администратором"
    await notify(
        bot, deal["seller_id"],
        f"✅ {who}. Вам выплачено <b>{fmt_amount(amount)} {deal['asset']}</b> "
        f"по сделке #{deal['id']} (комиссия сервиса {config.commission_percent}%).\n"
        "Средства зачислены на ваш баланс в @CryptoBot.",
    )
    await notify(
        bot, deal["buyer_id"],
        f"✅ Сделка #{deal['id']} завершена. Средства выплачены продавцу.",
    )
    deal = await db.get_deal(deal["id"])
    await ask_ratings(bot, deal)
    return True


async def refund_to_buyer(
    bot: Bot, db: Database, cp: CryptoPay, config: Config, deal, initiator: str
) -> bool:
    """Возврат покупателю из холда. initiator — 'seller' или 'admin'."""
    try:
        await cp.transfer(
            user_id=deal["buyer_id"],
            asset=deal["asset"],
            amount=deal["amount"],
            spend_id=f"refund_{deal['id']}",
            comment=f"Возврат по сделке #{deal['id']}",
        )
    except CryptoPayError as e:
        logger.error("Ошибка возврата по сделке %s: %s", deal["id"], e)
        for admin_id in config.admin_ids:
            await notify(
                bot, admin_id,
                f"🚨 Ошибка возврата покупателю по сделке #{deal['id']}: {e.name}",
            )
        return False

    await db.set_status(deal["id"], d.REFUNDED)
    who = "Продавец отменил сделку" if initiator == "seller" else "Спор решён администратором"
    await notify(
        bot, deal["buyer_id"],
        f"↩️ {who}. Вам возвращено <b>{fmt_amount(deal['amount'])} {deal['asset']}</b> "
        f"по сделке #{deal['id']}.\n"
        "Средства зачислены на ваш баланс в @CryptoBot.",
    )
    await notify(
        bot, deal["seller_id"],
        f"↩️ Сделка #{deal['id']} закрыта, средства возвращены покупателю.",
    )
    return True
